wallettool/src/main/java/org/bitcoinj/wallettool/WalletTool.java

```diff
@@ -892,23 +893,37 @@ private void send(PaymentSession session) {
             }
             setup();
             // No refund address specified, no user-specified memo field.
-            CompletableFuture<PaymentProtocol.Ack> future = session.sendPayment(ImmutableList.of(req.tx), null, null);
-            if (future == null) {
-                // No payment_url for submission so, broadcast and wait.
+            PaymentProtocol.Ack ack = session.sendPayment(ImmutableList.of(req.tx), null, null).get();
+            wallet.commitTx(req.tx);
+            System.out.println("Memo from server: " + ack.getMemo());
+        } catch (ExecutionException e) {
+            try {
+                throw e.getCause();
+            } catch (PaymentProtocolException.InvalidPaymentRequestURL e1) {
                 peerGroup.start();
-                peerGroup.broadcastTransaction(req.tx).future().get();
-            } else {
-                PaymentProtocol.Ack ack = future.get();
-                wallet.commitTx(req.tx);
-                System.out.println("Memo from server: " + ack.getMemo());
+                TransactionBroadcast broadcast = peerGroup.broadcastTransaction(req.tx);
+                try {
+                    // Wait for broadcast to be sent
+                    broadcast.future().get();
+                } catch (InterruptedException | ExecutionException be) {
+                    System.err.println("Failed to broadcast payment " + be.getMessage());
+                    System.exit(1);
+                }
```

Comment:
I think the broadcast via peerGroup code should be moved into a separate, private message. So, yes, I'm requesting changes in my own PR (although Github won't let me click the "Request Changes" button on my own PR)

Reply:
Yeah, give me 5-10 minute to change and force-push...

Reply:
> I think the broadcast via peerGroup code should be moved into a separate, private message. So, yes, I'm requesting changes in my own PR (although Github won't let me click the "Request Changes" button on my own PR)

It would be an improvement, yes. Like with the error handling improvement in the other PR. What I usually do – if possible – commit the improvements first (two commits in this case) and then rebase this PR on the improvements.

Reply:
OK, I'll try to create two new PRs, one with each of the two improvements.

Reply:
Turns out my advice doesn't apply to this situation, since the fallback code also changes śignificantly. Sorry for the noise.